Handle localized bundle names

tentenalex opened the pull request:
For apps that localize their bundle name in InfoPlist.strings, [[NSBundle mainBundle] infoDictionary] will not contain the key CFBundleName and the app crashes when trying to add the nil bundleName to the iosData dictionary. This fix checks whether the name is nil and, if it is, retrieves it from [[NSBundle mainBundle] localizedInfoDictionary] instead.

ezarowny commented:
@tentenalex would it be better to use [objectForInfoDictionaryKey](https://developer.apple.com/library/mac/documentation/Cocoa/Reference/Foundation/Classes/NSBundle_Class/#//apple_ref/occ/instm/NSBundle/objectForInfoDictionaryKey:) as it returns the localized value of a key if one is available?

ezarowny commented:
I'm having trouble reproducing this with a localized app. Can you give me some more information about how your app is set up?

* In ```App-Info.plist``` what do you have ```Bundle display name``` and ```Bundle name``` set to?
* What does your InfoPlist.strings file/files look like?

tentenalex commented:
@ezarowny Thanks for reminding me about that API!  This actually simplifies things a lot. 

I've added a commit using just -[NSBundle objectForInfoDictionaryKey:] instead of looking separately into infoDictionary and then localizedInfoDictionary. I'll clean up the other commit once you approve before merging.

Regarding our setup, I think the key point is that we have **completely removed** both CFBundleName and CFBundleDisplayName from App-Info.plist and added them to the respective InfoPlist.strings files as follows.  Maybe you cannot reproduce it because you have the base localization still directly in App-Info.plist?  

InfoPlist.strings (in Resources/en.lproj)
``` 
/* The short name of the app */
"CFBundleName" = "English Name";

/* The name of the app as displayed in the Springboard */
"CFBundleDisplayName" = "English Name";
```

InfoPlist.strings (in Resources/ja.lproj)
``` 
/* The short name of the app */
"CFBundleName" = "日本語のアプリ名";

/* The name of the app as displayed in the Springboard */
"CFBundleDisplayName" = "日本語のアプリ名";
```

tentenalex commented:
Incidentally, I just checked Apple's doc on the topic (https://developer.apple.com/library/ios/documentation/MacOSX/Conceptual/BPInternational/LocalizingYourApp/LocalizingYourApp.html) and the API they use there is: 

```smalltalk
NSString *appName = [[[NSBundle mainBundle] localizedInfoDictionary] objectForKey:@"CFBundleDisplayName"];
```

Not sure, maybe `-[NSBundle objectForInfoDictionaryKey:]` is a newer API, would have to check the headers for that.

ezarowny commented:
@tentenalex that's interesting! I've never tried to completely remove ```CFBundleName``` and ```CFBundleDisplayName``` from ```App-Info.plist```. I'll try that and see where it gets me for reproducibility.

ezarowny commented:
@tentenalex I was able to reproduce the crash when localization of ```CFBundleName``` and ```CFBundleDisplayName``` is done solely in ```InfoPlist.strings```.

Your changes work well and, as a bonus, they're an improvement on the existing code. Approved!

I'll merge and release after you cleanup the other commit. :+1:

ezarowny commented:
For what it's worth, I'm not sure that you have to remove any keys from ```App-Info.plist```. Localization should work fine without having to do that.

tentenalex commented:
@ezarowny, I cannot remember our reasoning for completely removing the keys from `App-Info.plist`, think it was for symmetry: if the Japanese name was defined in the strings file, so should the English one.    Maybe one can keep the ones in `App-Info.plist` and consider them the Base Localization, however, guess this fix is still relevant.  I now squashed the two commits and updated the commit message.
